open_spiel/observer.cc

```diff
@@ -118,15 +118,26 @@ std::shared_ptr<Observer> Game::MakeObserver(
 
   // Perfect information games provide public information regardless
   // of requested PrivateInfoType (as they have no private information).
-  if (GetType().information == GameType::Information::kPerfectInformation &&
-      !iig_obs_type->perfect_recall &&
-      (game_type_.provides_observation_tensor ||
-       game_type_.provides_observation_string)) {
-    if (iig_obs_type->public_info) {
-      return absl::make_unique<DefaultObserver>(*this);
-    } else {
+  if (GetType().information == GameType::Information::kPerfectInformation) {
+    // Handle the dummy case, where we do not use any public information.
+    // The game will just have empty private observations.
+    if (!iig_obs_type->public_info) {
       return absl::make_unique<NoPrivateObserver>(*this);
     }
+
+    // Distinguish two variants:
+    // Provide observer with or without perfect recall.
+    SPIEL_CHECK_TRUE(iig_obs_type->public_info);
```

Comment:
Remove this. Perhaps use an else clause.

Reply:
Actually maybe this logic can be further simplified.

const bool perfect_info_game = GetType().information == GameType::Information::kPerfectInformation;
  if (perfect_info_game && !iig_obs_type->public_info) {
      return absl::make_unique<NoPrivateObserver>(*this);
 }

if (iig_obs_type->perfect_recall && (game_type_.provides_information_state_tensor || game_type_.provides_information_state_string)) {
  SPIEL_CHECK(perfect_info_game || iig_obs_type->private_info == PrivateInfoType::kSinglePlayer);
  return absl::make_unique<InformationStateObserver>(*this);
}

/ same for default observer

Reply:
I simplified the logic.